src/main/java/io/cdap/plugin/ariba/source/AribaServices.java

```diff
@@ -207,26 +336,116 @@ private String getBase64EncodedValue() {
-      JsonNode jsonNode = objectMapper.readTree(responseStream);
+  public List<AribaColumnMetadata> getMetadata(String accessToken) throws
+    IOException, AribaException, InterruptedException {
+    HttpUrl.Builder templateBuilder = metadataTemplateBuilder(false, null);
+    AribaResponseContainer responseContainer = fetchAribaResponse(templateBuilder.build().url(), accessToken);
+    InputStream responseStream = responseContainer.getResponseBody();
+    JsonNode jsonNode = objectMapper.readTree(responseStream);
+    if (responseContainer.getHttpStatusCode() == HttpURLConnection.HTTP_OK) {
+
+      AribaMetaResponse res = gson.fromJson(jsonNode.toString(), AribaMetaResponse.class);
+      // Simple Fields
+      List<AribaColumnMetadata> columnDetails =
+        generateColumnValues(Properties.getNonObjectFields(res.getProperties()));
+      // Object Fields
+      columnDetails.addAll(getObjectFields(Properties.getObjectFields(res.getProperties())));
+      // Array Fields
+      columnDetails.addAll(getArrayFields(Properties.getArrayFields(res.getProperties())));
+      return columnDetails;
+    }
+    String errMsg = jsonNode.get(ResourceConstants.MESSAGE).asText() != null
+      ? jsonNode.get(ResourceConstants.MESSAGE).asText() :
+      ResourceConstants.ERR_NOT_FOUND.getMsgForKey();
+    throw new AribaException(errMsg, responseContainer.getHttpStatusCode());
+  }
+
+  /**
+   * @param aribaPluginConfig ariba plugin config
+   * @return JsonNode
+   */
+  public JsonNode createJob(AribaPluginConfig aribaPluginConfig, String pageToken) throws
+    AribaException, IOException, InterruptedException {
+    Request req = buildJobRequest(jobBuilder(pageToken).build().url(), aribaPluginConfig);
+    Response response;
+    do {
+      response = executeRequest(req);
+      if (response.code() == ResourceConstants.API_LIMIT_EXCEED) {
+        checkAndThrowException(response);
+      }
+    } while (response.code() == ResourceConstants.API_LIMIT_EXCEED);
+
+    AribaResponseContainer responseContainer = tokenResponse(response);
+    InputStream responseStream = responseContainer.getResponseBody();
+    if (responseContainer.getHttpStatusCode() == HttpURLConnection.HTTP_OK) {
+      availableLimit = Integer.parseInt(Objects.requireNonNull(response.header(RATE_LIMIT_DAY)));
+      LOG.info("Available limit from Create Job API: {}", availableLimit);
+      return objectMapper.readTree(responseStream);
+    }
+    throw new AribaException(response.message(), response.code());
+  }
+
+  /**
+   * @param accessToken access token
+   * @param jobId       job Id
+   * @return JsonNode
+   */
+  public JsonNode fetchJobStatus(String accessToken, String jobId)
+    throws IOException, AribaException, InterruptedException {
+    URL url = fetchDataBuilder(jobId).build().url();
+    Request req = buildFetchRequest(url, accessToken);
+    try (Response response = executeRequest(req)) {
+      do {
+        checkAndThrowException(response);
```

Comment:
I have added a max retry check but it was not required as in case the daily limit has reached then we are throwing exception and in other case we are just waiting for limit to be restored. And this was only happening when we are getting rate limit exception not any other type of exceptions.

Reply:
Fixed